### ISSUE TITLE
torch.nn.InstanceNorm2d CPU/GPU Inconsistency

### ISSUE THREAD

**Issue by ChaitanyaRS06** [pytorch/pytorch#159309]
### 🐛 Describe the bug

The torch.nn.InstanceNorm2d operation produces slightly different numerical results between CPU and GPU devices when using float16 dtype. The differences are small but consistent across multiple values in the output tensor.

```
import torch

# Create input tensor with explicit dtype and specific values
input_tensor = torch.tensor([[[[1.2, 2.3], [3.4, 4.5]], [[5.6, 6.7], [7.8, 8.9]]]], dtype=torch.float16)
instance_norm = torch.nn.InstanceNorm2d(2)
output = instance_norm(input_tensor)

input_tensor_gpu = torch.tensor([[[[1.2, 2.3], [3.4, 4.5]], [[5.6, 6.7], [7.8, 8.9]]]], dtype=torch.float16).to('cuda')
instance_norm_gpu = torch.nn.InstanceNorm2d(2)
output_gpu = instance_norm_gpu(input_tensor_gpu)

print('cpu: ', output)
print('gpu: ', output_gpu)


```

Output:

```
cpu:  tensor([[[[-1.3408, -0.4463],
          [ 0.4478,  1.3418]],

         [[-1.3408, -0.4480],
          [ 0.4480,  1.3408]]]], dtype=torch.float16)
gpu:  tensor([[[[-1.3418, -0.4468],
          [ 0.4473,  1.3418]],

         [[-1.3418, -0.4482],
          [ 0.4482,  1.3418]]]], device='cuda:0', dtype=torch.float16)
```

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Rocky Linux release 8.9 (Green Obsidian) (x86_64)
GCC version: (GCC) 8.5.0 20210514 (Red Hat 8.5.0-20)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.28

Python version: 3.11.6 | packaged by conda-forge | (main, Oct 3 2023, 10:40:35) [GCC 12.3.0] (64-bit runtime)
Python platform: Linux-4.18.0-513.18.1.el8_9.x86_64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-80GB
GPU 1: NVIDIA A100-SXM4-80GB
GPU 2: NVIDIA A100-SXM4-80GB

Nvidia driver version: 550.54.14
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Byte Order: Little Endian
CPU(s): 128
On-line CPU(s) list: 0-127
Thread(s) per core: 1
Core(s) per socket: 64
Socket(s): 2
NUMA node(s): 8
Vendor ID: AuthenticAMD
CPU family: 23
Model: 49
Model name: AMD EPYC 7742 64-Core Processor
Stepping: 0
CPU MHz: 2250.000
CPU max MHz: 2250.0000
CPU min MHz: 1500.0000
BogoMIPS: 4491.48
Virtualization: AMD-V
L1d cache: 32K
L1i cache: 32K
L2 cache: 512K
L3 cache: 16384K
NUMA node0 CPU(s): 0-15
NUMA node1 CPU(s): 16-31
NUMA node2 CPU(s): 32-47
NUMA node3 CPU(s): 48-63
NUMA node4 CPU(s): 64-79
NUMA node5 CPU(s): 80-95
NUMA node6 CPU(s): 96-111
NUMA node7 CPU(s): 112-127
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] pytorch-fid==0.3.0
[pip3] torch==2.6.0
[pip3] torchaudio==2.6.0
[pip3] torchmetrics==1.6.0
[pip3] torchvision==0.15.2+cu118
[pip3] triton==3.2.0
[conda] numpy

cc @ptrblck @msaroufim @eqy @jerryzh168

**Comment by HDCharles** [comment]
Its not guaranteed that they're exactly the same, see https://docs.pytorch.org/docs/stable/notes/numerical_accuracy.html

**Comment by malfet** [comment]
Re-opening an adding need reproduction: they don't have to be identical, but would be good to understand why it happens on such a small tensor, as op_math for both should be float32...

**Comment by eqy** [comment]
I think the native `batch_norm` that is used under the hood here does accumulation in `scalar_t`...
https://github.com/pytorch/pytorch/blob/750348b579dfc3be68b40e9883f07c6ab3307cd3/aten/src/ATen/native/cuda/Normalization.cu#L113
As I see
`batch_norm_transform_input_kernel` in the profile and this seems to be the only codepath that calls into it

**Comment by ngimel** [comment]
It would still accumulate in fp32 https://github.com/pytorch/pytorch/blob/b1bb860d3c812371b89a9725407230216e7369b5/aten/src/ATen/native/cuda/Normalization.cuh#L697
